test/forward_backward_compatibility/check_public_API_tests.py

```diff
@@ -0,0 +1,102 @@
+import os
+import subprocess
+import sys
+import tempfile
+
+
+BAD_PUBLIC_FUNC = b"""
+def new_public_func():
+    pass
+
+# valid public API functions have __module__ set correctly
+new_public_func.__module__ = None
+"""
+
+# Step 1. Make sure the public API test "test_correct_module_names" fails when a new file
+# introduces an invalid public API function.
+with tempfile.NamedTemporaryFile(dir="torch", suffix=".py") as f:
+    # Create a bad file underneath torch/ that introduces a new public API function incorrectly.
+    f.write(BAD_PUBLIC_FUNC)
+    f.seek(0)
+    mname = os.path.basename(f.name).split(".")[0]
+    breaking_name = f"torch.{mname}.new_public_func"
+    print(f"Creating an invalid public API function: {breaking_name}...")
+
+    # Run the public API test and make sure it fails.
+    res = subprocess.run(
+        [
+            sys.executable,
+            "test/test_public_bindings.py",
+            "-k",
+            "test_correct_module_names",
+        ],
+        capture_output=True,
+        text=True,
+        check=False,
+    )
+
+    # Fail the check if the public API test succeeded after introducing a bad file.
+    print("Running public API test 'test_correct_module_names'...")
+    if res.returncode == 0:
+        print(
+            "Expected the public API test 'test_correct_module_names' to fail after introducing "
+            "a bad file, but it succeeded! Check test/test_public_bindings.py for any changes "
+            "that may have broken the test."
+        )
+        sys.exit(1)
+
+    if breaking_name not in res.stderr:
+        print(
+            "Expected the public API test 'test_correct_module_names' to identify an invalid "
+            "public API, but it didn't! It's possible the test may not have run. Check "
+            "test/test_public_bindings.py for any changes that may have broken the test."
+        )
+        sys.exit(1)
+
+    print(
+        "Success! 'test_correct_module_names' identified a new, invalid public API function."
+    )
+
+# Step 2. Make sure that the public API test "test_modules_can_be_imported" fails when an existing
+# file is modified to introduce an invalid public API function.
```

Comment:
Huh, this test checks that "import your.module" doesn't crash :p 
What you describe here would be running test_correct_module_names again but after modifying an existing file right?

Reply:
Hm I'm trying to follow the description for this task:
```
In particular we want to make sure that both these tests run and are not broken (this happened before during some CI infra update). These tests can be (and have been) silently broken/disabled before and we have no signal about it! A way to do this is to have the lint shard run the following: 1. create a bad file 2. make sure test_public_bindings.py -k test_correct_module_names fails 3. append bad API to existing file 4. make sure test_modules_can_be_imported (same file) fails 5. undo the changes.
```

In practice, I've seen that modifying `torch/nn/functional.py` with an invalid public function does indeed fail `test_modules_can_be_imported`, but do we care about testing this test and, if so, if this the right way to do it?

Reply:
Updated to include the following test cases:
1. test_correct_module_names() fails for an invalid public API function in a new file
2. test_correct_module_names() fails for an invalid public API function in an existing file
3. test_modules_can_be_imported() fails when some module can't be imported